homeassistant/components/ondilo_ico/sensor.py

```diff
@@ -0,0 +1,186 @@
+"""Platform for sensor integration."""
+from datetime import timedelta
+import logging
+
+from ondilo import OndiloError
+
+from homeassistant.const import (
+    CONCENTRATION_PARTS_PER_MILLION,
+    DEVICE_CLASS_BATTERY,
+    DEVICE_CLASS_SIGNAL_STRENGTH,
+    DEVICE_CLASS_TEMPERATURE,
+    PERCENTAGE,
+    TEMP_CELSIUS,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+SENSOR_TYPES = {
+    "temperature": [
+        "Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        DEVICE_CLASS_TEMPERATURE,
+    ],
+    "orp": ["Oxydo Reduction Potential", "mV", "mdi:pool", None],
+    "ph": ["pH", "", "mdi:pool", None],
+    "tds": ["TDS", CONCENTRATION_PARTS_PER_MILLION, "mdi:pool", None],
+    "battery": ["Battery", PERCENTAGE, "mdi:battery", DEVICE_CLASS_BATTERY],
+    "rssi": [
+        "RSSI",
+        PERCENTAGE,
+        "mdi:wifi-strength-2",
+        DEVICE_CLASS_SIGNAL_STRENGTH,
+    ],
+    "salt": ["Salt", "mg/L", "mdi:pool", None],
+}
+
+SCAN_INTERVAL = timedelta(hours=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Ondilo ICO sensors."""
+
+    api = hass.data[DOMAIN][entry.entry_id]
+
+    async def async_update_data():
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables
+        so entities can quickly look up their data.
+        """
+        try:
+            pools = await hass.async_add_executor_job(api.get_pools)
+            for pool in pools:
+                pool["ICO"] = await hass.async_add_executor_job(
+                    api.get_ICO_details, pool["id"]
+                )
+                pool["devices"] = await hass.async_add_executor_job(
+                    api.get_last_pool_measures, pool["id"]
+                )
+
+            return pools
+
+        except OndiloError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="sensor",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=SCAN_INTERVAL,
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for devidx, _ in enumerate(pool["devices"]):
```

Comment:
Optional: Consider if its worth being defensive about ensuring "data_type" is found in SENSOR_TYPES

Reply:
Alternatively: iterate over every supported SENSOR_TYPE and check for it as a measure that exists in the set of measures, then create the entities just for those.  (This would only work if you did the "id lookup" pattern below and made the sensor data indexed by measure type)

Reply:
Done, thanks.